### PR TITLE
client, api: handle insufficient space error

### PR DESCRIPTION
**Pull request by stolowski**
Handle insufficient disk space errors in the api and client. For remove the error looks like this:

```
$ snap remove hello-world
error: cannot remove "hello-world" due to low disk space for automatic
       snapshot, use --purge to avoid creating a snapshot
```

(end-to-end spread will be included in #9221)

### CODE REVIEW

File: daemon/response.go [hunk 1]
```diff
@@ -473,6 +473,24 @@ func SnapChangeConflict(cce *snapstate.ChangeConflictError) Response {
...
+		"change-kind": dserr.ChangeKind,
+	}
+	return &resp{
+		Type: ResponseTypeError,
+		Result: &errorResult{
+			Message: dserr.Error(),
+			Kind: client.ErrorKindInsufficientSpace,
+			Value: value,
+		},
+		Status: 409,
```
Comment (stolowski) [inline daemon/response.go:490]: I suppose that should be 412?

Comment (pedronis) [reply daemon/response.go:490]: no, that one has a very precise meaning.  There is no great code for our case. See the discussion here for example:

https://evertpot.com/http/507-insufficient-storage

Comment (stolowski) [reply daemon/response.go:490]: Ok, sounds like 507 it is then for the lack of better candidate? 413 mentioned there doesn't really fit.

Comment (mvo5) [reply daemon/response.go:490]: Given that none really matches, should we just make up a 4XX error code?

File: cmd/snap/error.go [hunk 2]
```diff
@@ -222,6 +222,31 @@ If you understand and want to proceed repeat the command including --classic.
...
+			for i := range snaps {
+				snapNames[i] = fmt.Sprintf("%q", snaps[i])
+			}
+			names := strings.Join(snapNames, ",")
+			switch changeKind {
+			case "remove":
+				msg = fmt.Sprintf(i18n.G("cannot remove %s due to low disk space for automatic snapshot, use --purge to avoid creating a snapshot"), names)
+			case "install":
+				msg = fmt.Sprintf(i18n.G("cannot install %s due to low disk space"), names)
+			case "refresh":
```
Comment (mvo5) [inline cmd/snap/error.go:242]: (nitpick) this case is not tested AFAICT. maybe easy to add?

**Review by mvo5 — COMMENTED** [review]
This looks good to me. The fact that there is no matching error code is unfortunate. I wonder if we can just make one up (afaik e.g. twitter also does that in their api?).

File: cmd/snap/error.go (hunk 2 above)
Comment (stolowski) [reply cmd/snap/error.go:242]: Added (note, we won't see this in practice yet, needs refresh impl).

**Review by mvo5 — APPROVED** [review]
Approving as everything (beside the error code) is ready. Not sure how much we should block this on this - it seems there is no good option and if there is no good option we should just pick one of least bad ones and move on IMO.

File: client/errors.go [hunk 3]
```diff
@@ -129,6 +129,9 @@ const (
 
 	// ErrorKindDNSFailure: DNS not responding.
 	ErrorKindDNSFailure ErrorKind = "dns-failure"
+
+	// ErrorKindInsufficientSpace: not enough disk space to perform the request
+	ErrorKindInsufficientSpace ErrorKind = "insufficient-space"
```
Comment (pedronis) [inline client/errors.go:134]: I would expand to DiskSpace and disk-space for claritiy

File: client/errors.go [hunk 4]
```diff
@@ -129,6 +129,9 @@ const (
 
 	// ErrorKindDNSFailure: DNS not responding.
 	ErrorKindDNSFailure ErrorKind = "dns-failure"
+
+	// ErrorKindInsufficientSpace: not enough disk space to perform the request
```
Comment (pedronis) [inline client/errors.go:133]: this needs a dot at the end

File: cmd/snap/error.go [hunk 5]
```diff
@@ -222,6 +222,31 @@ If you understand and want to proceed repeat the command including --classic.
...
+		usesSnapName = false
+		values, ok := err.Value.(map[string]interface{})
+		if ok {
+			changeKind, _ := values["change-kind"].(string)
+			snaps, _ := values["snap-names"].([]interface{})
+			snapNames := make([]string, len(snaps))
+			for i := range snaps {
+				snapNames[i] = fmt.Sprintf("%q", snaps[i])
+			}
+			names := strings.Join(snapNames, ",")
```
Comment (pedronis) [inline cmd/snap/error.go:236]: there's strutil.Quoted that does what the loop+Join does I think

File: daemon/response.go [hunk 6]
```diff
@@ -473,6 +473,24 @@ func SnapChangeConflict(cce *snapstate.ChangeConflictError) Response {
 	}
 }
 
+// InsufficientSpace is an error responder used when an operation cannot
+// be performed due to low disk space.
+func InsufficientSpace(dserr *snapstate.InsufficientSpaceError) Response {
+	value := map[string]interface{}{
```
Comment (pedronis) [inline daemon/response.go:479]: we might want code like ChangeConflict that adds keys only if they are not ""

File: daemon/response.go (hunk 1 above)
Comment (pedronis) [reply daemon/response.go:490]: we'll have to document all of this for 2.48, and things wil be behind a feature flag for a bit, let's go with 507 for now and see what feedback we get

**Review by pedronis — COMMENTED** [review]
thanks, some detail comments

File: cmd/snap/error.go (hunk 5 above)
Comment (stolowski) [reply cmd/snap/error.go:236]: Indeed, thank you!

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**